File: spring_julia/spring.jl
using CSV

function dist(x,y)
    (x-y)^2
end

function get_min(m₀, m₁, m₂, i, j)
    if m₀ < m₁
        m₀ < m₂ && return i-1, j, m₀
        return i-1, j-1, m₂
    else
        m₁ < m₂ && return i, j-1, m₁
        return i-1, j-1, m₂
    end
end

function spring(x, y, ϵ)
    pathes = []

    Tx = length(x)
    Ty = length(y)

    C = zeros(Float64, (Tx, Ty))
    B = ones(Int64, (Tx, Ty, 2))
    S = ones(Int64, (Tx, Ty))

    C[1,1] = dist(x[1], y[1])

    for j in 2:Ty
        C[1, j] = C[1, j-1] + dist(x[1], y[j])
        S[1, j] = 1
        B[1, j, :] = [1, j-1]
    end

    for i in 2:Tx
        C[i, 1] = dist(x[i], y[1])
        S[i, 1] = i
        B[i, 1, :] = [1, 1]

        for j in 2:Ty
            pi, pj, m = get_min(C[i-1, j],
                                C[i, j-1],
                                C[i-1, j-1],
                                i, j)
            C[i,j] = dist(x[i], y[j]) + m
            B[i,j,:] = [pi, pj]
            S[i,j] = S[pi, pj]
        end

        imin = argmin(C[1:i, end])
        dmin = C[imin, end]

        dmin > ϵ && continue

        for j in 2:Ty
            C[i,j] < dmin && S[i,j] < imin && @goto yet
        end

        path = []
        push!(path, [imin, Ty])
        tempᵢ = imin
        tempⱼ = Ty

        while (B[tempᵢ, tempⱼ, 1] != 1 || B[tempᵢ, tempⱼ, 2] != 1)
            push!(path, B[tempᵢ, tempⱼ, :])
            tempᵢ, tempⱼ = B[tempᵢ, tempⱼ, :]
        end

        C[S .<= imin] .= Inf
        push!(pathes, path)
        @label yet
    end
    return pathes, C
end


function main()
    df = CSV.read("./data.csv", header=false, delim=",")
    data = df[!, 2]
    x = data[1:4:1000]
    y = data[1000:4:end]
    ϵ = 80

    pathes, C = spring(x, y, ϵ)
    println(pathes)
end

main()

File: spring/spring.py
import numpy as np


def dist(x, y):
    return (x - y)**2


def get_min(m0, m1, m2, i, j):
    if m0 < m1:
        if m0 < m2:
            return i - 1, j, m0
        else:
            return i - 1, j - 1, m2
    else:
        if m1 < m2:
            return i, j - 1, m1
        else:
            return i - 1, j - 1, m2


def spring(x, y, epsilon):
    Tx = len(x)
    Ty = len(y)

    C = np.zeros((Tx, Ty))
    B = np.zeros((Tx, Ty, 2), int)
    S = np.zeros((Tx, Ty), int)

    C[0, 0] = dist(x[0], y[0])

    for j in range(1, Ty):
        C[0, j] = C[0, j - 1] + dist(x[0], y[j])
        B[0, j] = [0, j - 1]
        S[0, j] = S[0, j - 1]

    for i in range(1, Tx):
        C[i, 0] = dist(x[i], y[0])
        B[i, 0] = [0, 0]
        S[i, 0] = i

        for j in range(1, Ty):
            pi, pj, m = get_min(C[i - 1, j],
                                C[i, j - 1],
                                C[i - 1, j - 1],
                                i, j)
            C[i, j] = dist(x[i], y[j]) + m
            B[i, j] = [pi, pj]
            S[i, j] = S[pi, pj]

        imin = np.argmin(C[:(i + 1), -1])
        dmin = C[imin, -1]

        if dmin > epsilon:
            continue

        for j in range(1, Ty):
            if (C[i, j] < dmin) and (S[i, j] < imin):
                break
        else:
            path = [[imin, Ty - 1]]
            temp_i = imin
            temp_j = Ty - 1

            while (B[temp_i, temp_j][0] != 0 or B[temp_i, temp_j][1] != 0):
                path.append(B[temp_i, temp_j])
                temp_i, temp_j = B[temp_i, temp_j].astype(int)

            C[S <= imin] = 100000000
            yield np.array(path), dmin


if __name__ == '__main__':

    T = 500
    t = .4

    x = np.sin(np.array(range(T)) / 10)
    y = np.sin((np.array(range(T)) / 10 + t * np.pi))
